typescript/packages/common-runner/src/builtins/fetch-data.ts

```diff
@@ -0,0 +1,78 @@
+import { type Node } from "@commontools/common-builder";
+import { cell, CellImpl, ReactivityLog } from "../cell.js";
+import { sendValueToBinding, findAllAliasedCells } from "../utils.js";
+import { schedule, Action } from "../scheduler.js";
+import { mapBindingsToCell } from "../utils.js";
+
+/**
+ * Fetch data from a URL.
+ *
+ * Returns the fetched result as `result`. `pending` is true while a request is pending.
+ *
+ * @param url - A cell containing the URL to fetch data from.
+ * @param schema - A cell containing the JSON schema to validate the fetched data.
+ * @returns { pending: boolean, result: any, error: any } - As individual cells, representing `pending` state, final `result`, and any `error`.
+ */
+export function fetchData(
+  recipeCell: CellImpl<any>,
+  { inputs, outputs }: Node
+) {
+  const inputBindings = mapBindingsToCell(inputs, recipeCell) as {
+    url: string;
+    result?: any;
+  };
+  const inputsCell = cell(inputBindings);
+
+  const pending = cell(false);
+  const fullResult = cell<any | undefined>(undefined);
+  const error = cell<any | undefined>(undefined);
+
+  const resultCell = cell({
+    pending,
+    result: fullResult,
+    error,
+  });
+
+  const outputBindings = mapBindingsToCell(outputs, recipeCell) as any[];
+  sendValueToBinding(recipeCell, outputBindings, resultCell);
+
+  let currentRun = 0;
```

Comment:
not sure what to do about this pattern - constantly repeating it seems bad?

Reply:
agreed. i was just going to add a couple of these with a lot of copy & paste until it becomes clearer which parts are shareable. i imagine after a few of these we see a pattern for async modules.

for instance here the distinguishing bit is that the results are spread across several individual cells, that can be independently updated. the rest is the same.